File: tmp_node1_sync/rae-core/rae_core/interfaces/graph.py
"""Abstract graph store interface for RAE-core."""

from typing import Any, Protocol, runtime_checkable
from uuid import UUID


@runtime_checkable
class IGraphStore(Protocol):
    """Abstract interface for knowledge graph storage."""

    async def create_node(
        self,
        node_id: UUID,
        node_type: str,
        tenant_id: str,
        properties: dict[str, Any] | None = None,
    ) -> bool:
        """Create a graph node."""
        ...

    async def create_edge(
        self,
        source_id: UUID,
        target_id: UUID,
        edge_type: str,
        tenant_id: str,
        weight: float = 1.0,
        properties: dict[str, Any] | None = None,
    ) -> bool:
        """Create a graph edge."""
        ...

    async def get_neighbors(
        self,
        node_id: UUID,
        tenant_id: str,
        edge_type: str | None = None,
        direction: str = "both",
        max_depth: int = 1,
    ) -> list[UUID]:
        """Get neighboring nodes."""
        ...

    async def delete_node(self, node_id: UUID, tenant_id: str) -> bool:
        """Delete a node and its edges."""
        ...

    async def delete_edge(
        self,
        source_id: UUID,
        target_id: UUID,
        edge_type: str,
        tenant_id: str,
    ) -> bool:
        """Delete an edge."""
        ...

    async def shortest_path(
        self,
        source_id: UUID,
        target_id: UUID,
        tenant_id: str,
        max_depth: int = 5,
    ) -> list[UUID] | None:
        """Find shortest path between nodes."""
        ...

    async def get_subgraph(
        self, node_ids: list[UUID], tenant_id: str, include_edges: bool = True
    ) -> dict[str, Any]:
        """Extract a subgraph."""
        ...

File: rae-core-rust/src/traits/graph.rs
use async_trait::async_trait;
use uuid::Uuid;
use serde_json::Value;
use std::collections::HashMap;
use anyhow::Result;

#[async_trait]
pub trait IGraphStore: Send + Sync {
    async fn create_node(
        &self,
        node_id: Uuid,
        node_type: &str,
        tenant_id: &str,
        properties: Option<HashMap<String, Value>>,
    ) -> Result<bool>;

    async fn create_edge(
        &self,
        source_id: Uuid,
        target_id: Uuid,
        edge_type: &str,
        tenant_id: &str,
        weight: f32,
        properties: Option<HashMap<String, Value>>,
    ) -> Result<bool>;

    async fn get_neighbors(
        &self,
        node_id: Uuid,
        tenant_id: &str,
        edge_type: Option<&str>,
        direction: &str,
        max_depth: usize,
    ) -> Result<Vec<Uuid>>;

    async fn delete_node(&self, node_id: Uuid, tenant_id: &str) -> Result<bool>;

    async fn delete_edge(
        &self,
        source_id: Uuid,
        target_id: Uuid,
        edge_type: &str,
        tenant_id: &str,
    ) -> Result<bool>;

    async fn shortest_path(
        &self,
        source_id: Uuid,
        target_id: Uuid,
        tenant_id: &str,
        max_depth: usize,
    ) -> Result<Option<Vec<Uuid>>>;

    async fn get_subgraph(
        &self,
        node_ids: Vec<Uuid>,
        tenant_id: &str,
        include_edges: bool,
    ) -> Result<Value>;
}
